site/en/guide/data_performance.ipynb

```diff
@@ -0,0 +1,1142 @@
+    "#### Parallel mapping\n",
+    "\n",
+    "Now, use the same pre-processing method but applying it in parallel on multiple samples."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Execution time: 0.2288658139987092\n"
+     ]
+    }
+   ],
+   "source": [
+    "iterate(\n",
+    "    ArtificialDataset()\n",
+    "    .map(\n",
+    "        mapped_function,\n",
+    "        num_parallel_calls=tf.data.experimental.AUTOTUNE\n",
+    "    )\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![Parallel mapping](./parallel_map.svg)\n",
+    "\n",
+    "Now, you can see on the plot that the pre-processing steps can overlap allowing to reduce the overall time for a single iteration."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Vectorizing mapping\n",
+    "\n",
+    "Invoking the user-defined function passed into the `map` transformation has overhead related to scheduling and executing the user-defined function.\n",
+    "We recommend vectorizing the user-defined function (that is, have it operate over a batch of inputs at once) and apply the `batch` transformation _before_ the `map` transformation.\n",
+    "\n",
+    "To apply this change to our running example, define a new parsing function, acting on a batch of serialized examples, and call it after batching our dataset."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Execution time: 1.4374910180013103\n"
+     ]
+    }
+   ],
+   "source": [
+    "_batch_map_num_items = 25\n",
+    "def factorial(x):\n",
+    "    if x <= 1:\n",
+    "        return 1.0\n",
+    "    else:\n",
+    "        f = x * factorial(x - 1.0)\n",
+    "        return f\n",
+    "        \n",
+    "\n",
+    "iterate(\n",
+    "    ArtificialDataset(\n",
+    "        num_samples=_batch_map_num_items  # Generate more items\n",
+    "    ).map(  # Apply function one item at a time\n",
+    "        lambda x: factorial(8)\n",
+    "    ).batch(  # Batch and unbatch to compare with the vectorized version\n",
+    "        _batch_map_num_items, drop_remainder=True\n",
+    "    ).unbatch()\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![Scalar map](./scalar_map.svg)\n",
+    "\n",
+    "In this example, you can see that the mapped function is applied for each sample. While this function is very fast, it has some overhead that impact the time performance."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Execution time: 1.453609274001792\n"
```

Comment:
when I run this, I get the same time for vectorized vs non-vectorized mapping. To show the performance, I think we need a non-artificial dataset with a large batch size and very cheap map function.

Reply:
Yes, I a little bit stuck with this part. My idea would be to divide all values of the tensor by a value which is long to compute (as a recurcive factorial or whatever). However, from the results here, I suppose that the value is kept in some kind of cache, hence not taking so much time to be computed for "scalar" version...
In my first experiment, I used a `sleep(1ms)` for the scalar case and a `sleep(batch_size ms)` for the vectorized one, which gave good result but not applicable in graph mode...
Maybe playing with loops can provide a acceptable result.

Why do you think that a non-artificial dataset would better show performance improvements?

Reply:
Vectorization only has a noticeable impact when the overhead of scheduling map functions is large compared to the time it takes to run the map function. Scheduling a map function takes around 10 microseconds, so to see an impact we need to be iterating very fast, which isn't really possible when using the sleep-based artificial dataset.

We can do something like this:

```python
import tensorflow as tf
import time
tf.enable_eager_execution()

elements = 10000
batch_size = 1024
num_batches = elements/batch_size
ds = tf.data.Dataset.range(elements)

def f(x):
  return x+1

def vectorized_f(batch):
  return tf.map_fn(f, batch)

non_vectorized = ds.map(f).batch(batch_size)
vectorized = ds.batch(batch_size).map(vectorized_f)

def benchmark(dataset):
  start = time.time()
  # iterate using skip so that all iteration is done in c++.
  next(iter(dataset.skip(num_batches-1)))
  return time.time() - start

# warmup
benchmark(non_vectorized)
benchmark(vectorized)

print("Non-vectorized takes " + str(benchmark(non_vectorized)) + " seconds")
print("Vectorized takes " + str(benchmark(vectorized)) + " seconds")

# Non-vectorized takes 0.550505876541 seconds
# Vectorized takes 0.166946172714 seconds
```

Reply:
Wow, this example exhibits an impresive 30% performance improvements! But it requires:
- to introduce a new dataset − even if easy to use, it creates a special case in the guide;
- to write a new benchmarking loop with special iteration mechanisms.

However, I did not manage to provide any example which work well with both, eager (`py_function`) and graph execution. I even try to use shorter `sleep` for dataset but, results are still unpredictable and not reproducible, as well as making the images less clear because "hidden" times (start of an epoch, various logic) are more/too visible compared to the measured steps...

So, I am a little bit stuck for the moment. If we choose an example similar to yours, maybe we have to introduce it more thoroughly, later in the guide (after "Caching").

A more general question. Is it required to use `tf.map_fn`? The `add` method already applies to all elements of the batch (thanks to broadcasting from what I understand).

Reply:
Here's a simplified example that shows over 10x speedup by using `tf.add` as you suggested. We see speedup even without introducing any special iteration mechanism. I think it's ok to introduce a new (very simple) dataset. I agree it should go at the end of the guide after we've gone over all the optimizations that apply to ArtificialDataset.

```
def benchmark(dataset):
  start = time.time()
  for elem in dataset:
    pass
  return time.time() - start

ds = tf.data.Dataset.range(10000)
non_vectorized = ds.map(lambda x: x+1).batch(256)
vectorized = ds.batch(256).map(lambda x: tf.add(x, 1))

print("Vectorized takes " + str(benchmark(vectorized)) + " seconds")
print("Non-vectorized takes " + str(benchmark(non_vectorized)) + " seconds")
```

Reply:
Updated with your example. I used `lambda x: x+1` for both examples, `tf.add` is not part of the TF2.0 API and make the syntax complicated.
Interestingly, adding the `unbatch` transformation afterward makes both datasets roughly on par.